### ISSUE TITLE
"ndim" argument to flatten

### ISSUE THREAD

**Issue by petered** [numpy/numpy#8102]
Theano's [tensor.flatten](http://deeplearning.net/software/theano/library/tensor/basic.html#theano.tensor._tensor_py_operators.flatten) function takes an optional `ndim` argument, specifying the dimension from which you'd like to flatten.

Eg, if arr has shape (3, 4, 2)
`barr = arr.flatten(1)`
Will result in barr having shape (3, 8)

It would be very nice for numpy's flatten to have the same syntax, as it would allow for some nice one liners that currently have to be split up.  For example:

```
images = get_images()  # (n_images, size_y, size_x, 3)
flattened_images = images.reshape(images.shape[0], -1)
```
Could be simplified to 

```
flattened_images = get_images.flatten(2)
```

**Comment by shoyer** [comment]
Flattening from a certain dimension onward is a really useful operation for neural networks, but I haven't seen it used in other domains.

NumPy's flatten and ravel already have an optional argument, so you would need to write this out in full form, e.g., `arr.flatten(ndim=1)`. Given how easy (and frankly more convenient) it is write such a helper function for yourself, I would be inclined to pass on this functionality for NumPy.

**Comment by petered** [comment]
I'd say it's a pretty generally common operation anywhere in machine learning or image processing.  Matlab's equivalent is `flattened_images = images(:, :)`.  Of course I could do it with a helper, but it would be nice to have code that's portable.

An alternative would be to allow `None` as an argument to reshape to indicate "keep current dimension", eg `flattened_images = images.reshape(None, -1)`.

**Comment by mhvk** [comment]
As a single point of reference: I've written quite a few similar `reshape`s and feel it is a bit awkward to have to specify parts of the original shape. For quite a few, though, `-1` was not necessarily the other argument, so it would be nice to be more general. Representing a current shape with `None` would be an option, though I wish I could think of an option that would be a bit more obvious when seeing it.

More to the point of the original suggestion, another way to use `flatten` (and then also `ravel`, which doesn't make a copy)  that would be more general than using `ndim` would be to allow one to indicate which axes to flatten/ravel. So, one might write `images.ravel(axes=(1, 2))` (with a default of all axes). I think for readability it would be best if this argument was always spelled out, so from that perspective it is not a problem that there is already another argument to the method.

Actually, perhaps an `axes` argument to `reshape` might also work: that the new shape applies only to the axes specified.

But perhaps all this is too much trouble!

**Comment by petered** [comment]
Giving an "axes" argument to ravel or reshape seems a bit confusing, as I'm not sure what to expect from `images.ravel(axes=(1, 3))`.  (A `(n_images, size_x, size_y*3)` array I guess, though this would require moving the actual data around, which is not expected from ravel.)

Supposing we want to flatten out the pixels in an array of images, so that a `(n_images, size_y, size_x, 3)` array becomes `(n_images, size_y*size_x, 3)`.  

I think the best would be:
`flattened_ims = images.reshape(None, -1, None)`

Another would be:
`flattened_ims = images.ravel(start_axis=1, end_axis=2)`

One good reason for all this would be to avoid breaking up expressions that could otherwise be one-liners, as in the example on top.

**Comment by shoyer** [comment]
A generic placeholder like `None`/`keepdim` for existing dimension sizes *does* seem convenient (and would also work nicely in TensorFlow, where axis sizes can be unknown). My main concern is that it may not be entirely clear which original axes are being referred to, especially for trailing axes. I guess we could count `keepdim` markers after a `-1` from the end, sort of like trailing indexers after `...` in `__getitem__`? What then would `.reshape(None, size_x * size_y, None)` mean? Logically, that would error, but I think it would be confusing to break the existing invariant that -1 can be substituted for the remaining size.

I do like @mhvk's proposal of an explicit `axes` argument. Conversely, we could have an argument like `exclude`/`exclude_axes` for indicating axes not to be reshaped. I'm not sure that's really better, but though I'd put it out there anyways. Either way, I like specifying a list of `axes` explicitly much better than a start/end axis.

**Comment by petered** [comment]
Yes, that's what I was thinking.. count them from the end after the first -1 or numeric dimension (and throw an error if you have more None's than remaining dimensions, or if the user provides None's between numeric arguments). 

So, if images is shape `(n_samples, size_y, size_x, 3)`:

* `images.reshape(None, size_x * size_y, None)` would be `(n_samples, n_size_y*size_x, 3)`
* `images.reshape(None, -1, None)` would be `(n_samples, n_size_y*size_x, 3)`
* `images.reshape(None, -1)` would be `(n_samples, n_size_y*size_x*3)`
* `images.reshape(None, -1, None, None)` would be be unchanged
* `images.reshape(None, -1, None, None, None)` would cause an error
* `images.reshape(None, None, None, None)` would be unchanged
* `images.reshape(images.shape[0], None, -1)` would cause an error
* `images.reshape(-1, None, None)` would be `(n_images*size_y, size_x, 3)` (whyever you'd want that)
* `images.reshape(None, 46326324, None)` would probably cause an error

**Comment by shoyer** [comment]
So the rule would be that the new shape can only include contiguous blocks of `None` at the start and/or the end? This feels quite restrictive, given that only a small fraction of all syntactically valid statements will be valid operations.

If `None` intuitively means "use the existing size for a dimension" then cases like `images.reshape(n_samples, None, -1)` or `images.reshape(-1, None, None, 1)` giving errors is going to be surprising. If using `None` with reshape requires careful reading of the fine print to use correctly, then it's probably bad API design. I would rather add keyword arguments like `ndim_min` and `ndim_max` (but `axes` is probably less ambiguous).

**Comment by petered** [comment]
I can't see how axes would work with nonadjacent axes, maybe you could give an example?

I guess your option of `images.ravel(ndim_min=1, ndim_max=2)` could be replace `images.reshape(None, -1, None)`, but I find the former more ambiguous because it doesn't tell you anything about how many dimensions the resulting array has.

Another option would be an einsum-like notation where `images.reshape('s,i,j,c->s,ij,c')` behaves like `images.reshape(images.shape[0], -1, images.shape[-1])`, and also has the advantage of asserting that the initial dimensionality of image is as expected.

**Comment by shoyer** [comment]
You're right, I think all `axes` would indeed need to be adjacent for the reshape to have sensibly ordered outputs. So we have the limitation of being basically equivalent to `ndim_min`/`ndim_max` either way. Still, I think `images.ravel(axes=[1, 2])` or `images.reshape((-1,), axes=[1, 2])` is more readable/self-explanatory than `images.reshape(None, -1, None)`.

**Comment by eric-wieser** [comment]
I like the look of `images.reshape((-1,), axes=[1, 2])` (and the `ravel` that should come with it). Would this work in both directions, for inserting axes?

```python
image = np.zeros(W, H, 3)
pixels = images.reshape((-1,), axes=[0, 1])
image_again = pixels.reshape((W, -1), axes=[0])
```

I'm not sure how I feel about the `axes=[0]` there.

**Comment by shoyer** [comment]
> Would this work in both directions, for inserting axes?

Yes, I don't see why not.

> I'm not sure how I feel about the axes=[0] there.

We could probably also allow `axes=0` as a shortcut.

**Comment by eric-wieser** [comment]
Shouldn't this be `axis=[1,2]` or `axis=0` for consistency with `.sum`?

**Comment by shoyer** [comment]
Probably, yes.

**Comment by petered** [comment]
That could work.  But again, what do you suggest for non-adjacent axes?  eg. `pixels = images.reshape((-1,), axes=[0, 2])`  Should that just throw an exception?

**Comment by shoyer** [comment]
> But again, what do you suggest for non-adjacent axes? eg. pixels = images.reshape((-1,), axes=[0, 2]) Should that just throw an exception?

It's not even obvious to me what the shape of `images.reshape((-1,), axes=[0, 2])` should be, nevermind how the axes can be made to line up. For now, I would just raise -- we can always add support later if someone can figure out consistent rules.

Note that `reshape`, `flatten` and `ravel` are implemented in C, so that's where the fix would need to happen (search for `array_reshape`/`array_flatten`/`array_ravel`).

**Comment by eric-wieser** [comment]
If we have:

    images = np.zeros((A, B, C))

Then I would say the most sensible choice is:
    
    flat = images.reshape((-1,), axes=[0, 2])
    flat.shape == (A*C, B)
    images[i,j,k] = flat[i*C + k, j]

That is, bring all the collapsed axes into the place of the first one - this is sort of consistent with how advanced indexing behaves.

**Comment by shoyer** [comment]
@eric-wieser I agree, that makes sense. But what would `images.reshape((A, C), axes=[0, 2])` do? I think the only consistent choice is to insert all new axes at the first position, but `images.reshape((A, C), axes=[0, 2]).shape == (A, C, B)` might be a little surprising.

**Comment by seberg** [comment]
If you want silly ideas ;), check: https://gist.github.com/seberg/3430219 (the second function/ReIndex stuff)

What to do if the axes are non-contiguous seems about as sane as fancy indexing and we all know how many people understand that....

So an axes argument, why not, but I would enforce that there are no "holes".  You could argue that it is easier to just have the first and last axis. On the other hand, when you think of splitting up one axis for example, "axis" is much more straight forward.

**Comment by eric-wieser** [comment]
> I think the only consistent choice is to insert all new axes at the first position

I see two exclusive possible choices here:

* Only insert axes at the first position
* Only allow non-sequential reshapes to have the same input and output dimensionality

I think the second behaviour is clearer, but the first matches the design and saneness of fancy indexing

**Comment by seberg** [comment]
Btw. if someone goes so far and creates a new C-side function for this (which is necessary, though it may not have to be exposed), might want to also add a `copy={None, True, False}` parameter. Where None is "If necessary", False is "error if impossible to create view" and True forces the copy in any case.
